src/relay/QueryRenderer.js

```diff
@@ -18,25 +20,41 @@ type Props<RenderProps> = {|
   query: () => mixed,
   render: (RenderArgs<RenderProps>) => ?React$Element<*>,
   variables: { [string]: mixed },
+  logoutCallback?: () => void,
 |};
 
-const QueryRenderer = <RenderProps>(props: Props<RenderProps>) => (
-  <UserContext.Consumer>
-    {({ loginToken }: UserContextType) => (
-      <LanguageContext.Consumer>
-        {language => {
-          const locale = frontendLanguageToLocale[language];
-
-          return (
-            <OriginalQueryRenderer
-              {...props}
-              environment={createEnvironment(loginToken, locale)}
-            />
-          );
-        }}
-      </LanguageContext.Consumer>
-    )}
-  </UserContext.Consumer>
-);
+type ContextTypes = {
+  loginToken: ?string,
+  defaultLanguage: string,
+  onLogout: onLogout,
+};
+type RenderState<RenderProps> = {
+  props: ?RenderProps,
+  error: ?Error,
+};
+class QueryRenderer extends React.Component<ContextTypes & Props<RenderProps>> {
+  renderContainer = (renderState: RenderState<RenderProps>) => {
+    if (renderState.error) {
+      if (renderState.error.message === ERROR_FORBIDDEN) {
+        this.props.onLogout().then(() => {
```

Comment:
should use async/await instead

Reply:
How do we do that before they land [async rendering](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html) in React?

Reply:
good point 🤔 

maybe 
```
(async () => { 
  await something(); 
})().catch(console.error)
```